### ISSUE TITLE
Normalise licensing for projects in the IndieWeb GitHub org.

### ISSUE THREAD

**Issue by Zegnat** [indieweb/wiki#39]
I am opening this issue here because there is no other centralised repository for organisation issues. ([Asked about this on IRC.](https://chat.indieweb.org/2017-02-24#t1487946504219000))

---

Below is a table of all 28 current repositories in the IndieWeb GitHub organisation, the licence they are made available under, and links to issues that were opened because of the licensing.

Many of those issues were opened by @marclaporte on several of the Apache licensed PHP projects is because he ran into trouble including these libraries in other projects. He suggests re-licensing or dual-licensing the respective repositories with MIT, so it would allow him to include these libraries. Through @sebsel I found the issue on [indieweb/php-comments](https://github.com/indieweb/php-comments) and [asked why not make the move to CC0](https://github.com/indieweb/php-comments/issues/13#issuecomment-275389032)?

I would like to use this as a central issue to discuss wether it would be worth it to move **all** code in the IndieWeb GitHub organisation to the same licensing.

Personally, I would like to see a dual-licensed configuration of [CC0 1.0 Universal](https://tldrlegal.com/license/creative-commons-cc0-1.0-universal) and [MIT](https://tldrlegal.com/license/mit-license). Why dual-license something that gets pledged to the public domain? Because some [countries may not allow perpetualy signing away rights](http://softwareengineering.stackexchange.com/a/147120). Offering companies the easy out with a secondary licence might be the best thing to do, a bit like [SQLite selling licences](https://www.sqlite.org/copyright.html) (read their reasons for obtaining a licence).

There is an important precedent for [the move to CC0 on the php-mf2 repository](https://github.com/indieweb/php-mf2/issues/76).

What are people’s opinions on this?

---

| Repo                                                                                   | Licence                     | Licensing issues                                                     |
| -------------------------------------------------------------------------------------- | --------------------------- | --------------------------------------------------------------------:|
| [IndieWeb Week Website](https://github.com/indieweb/indiewebweek)                      | CC0 1.0 Universal‡          |                                                                      |
| [IndieAuth Client](https://github.com/indieweb/indieauth-client-php)                   | Apache License, Version 2.0 | [#9](https://github.com/indieweb/indieauth-client-php/issues/9)      |
| [php-mf2](https://github.com/indieweb/php-mf2)                                         | CC0 1.0 Universal           | **[#76](https://github.com/indieweb/php-mf2/issues/76)**             |
| [WordPress IndieWeb](https://github.com/indieweb/wordpress-indieweb)                   | MIT                         |                                                                      |
| [link-rel-parser-php](https://github.com/indieweb/link-rel-parser-php)                 | Apache License, Version 2.0 | [#5](https://github.com/indieweb/link-rel-parser-php/issues/5)       |
| [Comments Presentation](https://github.com/indieweb/php-comments)                      | Apache License, Version 2.0 | [#13](https://github.com/indieweb/php-comments/issues/13)            |
| [verify-me](https://github.com/indieweb/verify-me)                                     | Apache License, Version 2.0 |                                                                      |
| [Webmention Client](https://github.com/indieweb/mention-client-php) (PHP)              | Apache License, Version 2.0 | [#30](https://github.com/indieweb/mention-client-php/issues/30)      |
| [chat.indieweb.org](https://github.com/indieweb/chat.indieweb.org)                     | Apache License, Version 2.0 |                                                                      |
| [Webmention Client](https://github.com/indieweb/mention-client-ruby) (Ruby)            | Apache License, Version 2.0 |                                                                      |
| [IndieWebify.me](https://github.com/indieweb/indiewebify-me)                           | _none_                      |                                                                      |
| [2016.indieweb.org](https://github.com/indieweb/2016.indieweb.org)                     | _none_                      |                                                                      |
| [This Week in the IndieWeb](https://github.com/indieweb/this-week)                     | Apache License, Version 2.0 |                                                                      |
| [blank-gh-site](https://github.com/indieweb/blank-gh-site)                             | _none_                      |                                                                      |
| [rel-me](https://github.com/indieweb/rel-me)                                           | MIT*                        |                                                                      |
| [Microformats2 (ruby)](https://github.com/indieweb/microformats2-ruby)                 | CC0 1.0 Universal           |                                                                      |
| [indieweb.org](https://github.com/indieweb/indieweb.github.io)                         | CC0 1.0 Universal           |                                                                      |
| [Indie Web Camp branding](https://github.com/indieweb/branding)                        | CC0 1.0 Universal           |                                                                      |
| [php-mf2-shim](https://github.com/indieweb/php-mf2-shim)                               | BSD 3-Clause License†       |                                                                      |
| [LinkRelParser (Ruby)](https://github.com/indieweb/link_rel_parser-ruby)               | CC0 1.0 Universal           |                                                                      |
| [Representative H-Card Parsing](https://github.com/indieweb/representative-h-card-php) | Apache License, Version 2.0 | [#1](https://github.com/indieweb/representative-h-card-php/issues/1) |
| [Date Formatter](https://github.com/indieweb/date-formatter-php)                       | Apache License, Version 2.0 | [#6](https://github.com/indieweb/date-formatter-php/issues/6)        |
| [php-original-post-discovery](https://github.com/indieweb/php-original-post-discovery) | MIT                         |                                                                      | 
| [wiki](https://github.com/indieweb/wiki)                                               | Unlicense                   |                                                                      |
| [jf2 validator](https://github.com/indieweb/jf2-validator)                             | Apache License, Version 2.0 |                                                                      |
| [IndieWebCamp Assets](https://github.com/indieweb/assets)                              | _none_                      |                                                                      |
| [mynameisme.org](https://github.com/indieweb/mynameisme.org)                           | _none_                      |                                                                      |
| [newBase60py](https://github.com/indieweb/newBase60py)                                 | CC0 1.0 Universal           |                                                                      |

*: The licence had to be taken from `composer.json`.
†: Guessing 3-Clause, did not do a full text comparison.
‡: The repo is available under CC0, but all contributions are dual-licensed CC0 and OWFa 1.0.

Notes:

* php-mf2-shim talks mentions MIT in its `composer.json` but provides the BSD licence.

**Comment by marclaporte** [comment]
Hi Martijn!

Thank you for preparing this.

I agree it would be better to have a recommended license for all IndieWeb (sub)-projects.

If / once this license is agreed upon, we could ask all contributors to agree to this for new contributions. And then, we can take the time to go back to all previous contributors to ask if they can agree to the new license. This takes time, so the sooner we start, the better.

As a reference, here is the saga of the license change for Bootstrap:
https://github.com/twbs/bootstrap/issues/2054

Best regards,

**Comment by tantek** [comment]
I'd lean toward CC0 since that's consistent with the wiki(s) (including microformats.org), and could live with dual licensing with MIT since that seems to add something for some folks?

One nice feature of the CC0 / wiki-compat is that code / examples can be freely copy/pasted from the indieweb repos vs the wikis without worry.

**Comment by Zegnat** [comment]
> One nice feature of the CC0 / wiki-compat is that code / examples can be freely copy/pasted from the indieweb repos vs the wikis without worry.

That’s definitely an important point :+1:

Also, as I thought about this [during chat](https://chat.indieweb.org/dev/2017-03-01#t1488401861532000) and not when I was writing the issue above: it would be good to decide on the texts of LICENSE.md and CONTRIBUTING.md before opening requests on the other repos. Basically “a template to copy” (— [tantek](https://chat.indieweb.org/dev/2017-03-01#t1488401957195000)).

**Comment by petermolnar** [comment]
This might come handy as simple comparison between the most common licenses:
https://choosealicense.com/licenses/
Unfortunately it excludes CC.

**Comment by dg01d** [comment]
Looking over the table, it appears that the Apache Licence 2.0 is the most popular. Moving from that to MIT would mean losing the "Patent Protection" clause of the ASL:

> If You institute patent litigation against any entity (including a cross-claim or counterclaim in a lawsuit) alleging that the Work or a Contribution incorporated within the Work constitutes direct or contributory patent infringement, then any patent licenses granted to You under this License for that Work shall terminate as of the date such litigation is filed.

Its important to understand that this isn't a case of just changing some words - these are issues that people may have a fundamental belief in.

**Comment by Zegnat** [comment]
> This might come handy as simple comparison between the most common licenses:
> https://choosealicense.com/licenses/
> Unfortunately it excludes CC.

Creative Commons (and thus [CC0](https://choosealicense.com/licenses/cc0-1.0/)) is explicitly only listed under [Non-Software Licenses](https://choosealicense.com/non-software/) by choosealicense.com. But as [per the CC0 FAQ](https://wiki.creativecommons.org/wiki/CC0_FAQ#May_I_apply_CC0_to_computer_software.3F_If_so.2C_is_there_a_recommended_implementation.3F), it is fine to use CC0 for software even when other Creative Commons licences should not be used.

choosealicense.com seems to prefer [Unlicense](https://choosealicense.com/licenses/unlicense/) for the Public Domain dedication. While both Unlicense and CC0 are seen as “free software licenses” by the FSF, [they recommend the latter](https://www.gnu.org/licenses/license-list.en.html#Unlicense):

> If you want to release your work to the public domain, we recommend you use CC0. CC0 also provides a public domain dedication with a fallback license, and is more thorough and mature than the Unlicense.

> Looking over the table, it appears that the Apache License 2.0 is the most popular. Moving from that to MIT would mean losing the "Patent Protection" clause of the ALv2.0

Note that the main idea is moving to CC0, not MIT. Both to match the wiki and just to generally allow anyone to do anything with IndieWeb code.

The only reason I bring up MIT is because waiving all rights and dedicating something to the public domain is hard. It might even be impossible in some jurisdictions. And while CC0 tries to fix this by also being a licence next to being a waiver, people might feel more at ease if they can opt for a more accepted minimal licence. This might also be essential for people in countries like Norway, where copyright to public domain works goes to the state and [certain industries’ overseers can charge levies](https://github.com/unitedstates/licensing/issues/31).

This is basically copied from SQLite, who offer a commercial licence for those who do not want to rely on public domain code. Except it makes no sense for IndieWeb to start selling licences.

Does “Patent Protection” make any sense in code to which the creators (try to) waive their rights? Can you even still patent a specific solution you came up with, if that solution has already been dedicated to the Public Domain? I am tempted to say no, but obviously I am not an IP lawyer.

**Comment by dg01d** [comment]
IAAL&IANYL. That said: the CC0 licence specifically retains patent rights to the creator. There are different 'levels' of Intellectual Property, the CC0 treats those differently. It releases copyright  and related ownership rights to the Public Domain, it does not release patent & trademark rights. 
While the CC & FSF state that its fine for software, the OSI balked at calling it an open licence for just that reason.
Please understand that I'm not advocating against the CC0, just trying to establish the differences and what is and isn't assigned. 
The situation you posit is *exactly* what the CC0 was designed to do: release datasets to the public domain while retaining patent ownership of the methodologies to create and indeed extract the data from raw scientific inputs. It wasn't designed for software at all.

**Comment by voxpelli** [comment]
Just noticed that Google doesn't permit their employees to patch repositories with CC0 licenses, so unless a Google employee goes through the special process of claiming personal copyright for patches to such a project then they can't patch them: https://opensource.google.com/docs/patching/#stuff-you-cant-do

I don't know if other companies have similar policies, anyone knows?

Perhaps @willnorris can give some more insight or feedback into this and whether it becomes a problem for people working at eg. Google to have CC0 here?

**Comment by Zegnat** [comment]
That’s interesting. I think the give away is the first line of [the page about personal projects](https://opensource.google.com/docs/iarc/):

> As part of your employment agreement, Google most likely owns intellectual property (IP) you create while at the company.

With Google owning the rights to the code you write, it is easy to understand why they wouldn’t allow you personally dedicating the code to the Public Domain. (You wouldn’t even be able to do so, Google would have to do that for you.)

You are not allowed to sign CLAs either, which I expect is for the same reason: CLAs usually require you to assign your copyright to a company/organisation/foundation to make sure an author doesn’t go missing when something like a relicense comes up (think VLC), but if Google owns your rights you can’t personally sign them away.

Contributing patches to other places is probably OK because those involve licensing your rights, not giving them up to someone else (or to the public domain). Google will still own the code written.

Fun fact, Google themselves do [accept CC0 and Unlicense for use in their projects as “unencumbered”](https://opensource.google.com/docs/thirdparty/licenses/#unencumbered). (If you read on they address the issue with other public domain dedications, which I hope we addressed by allowing companies to use any code under MIT.)

**Comment by willnorris** [comment]
That's correct, Google's concern is primarily with patching CC0 projects, not with using them.  It has to do with restricting the ability to relicense the code, which is a concern unique to the language in CC0.

I'm increasingly becoming a fan of the [0-Clause BSD](https://spdx.org/licenses/0BSD.html) (aka the [Free Public License](https://opensource.org/licenses/FPL-1.0.0)) for cases like these where you really don't care about attribution.

**Comment by dg01d** [comment]
Yes, I quite agree. Note that the OSI calls this the ["Free Public Licence"](https://opensource.org/licenses/FPL-1.0.0)

Having been a long-standing MIT proponent, I've started using the FPL/0BSD instead.

**Comment by Zegnat** [comment]
The 0-Clause BSD is really interesting to me. I think I will be following @dg01d in using it instead of MIT.

We had [a discussion on the IndieWeb development channel](https://chat.indieweb.org/dev/2017-07-11#t1499793351878000) about the license and @dg01d was kind enough to share his notes on the difference between MIT and 0BSD. When reading the chat logs, please note that this was an informal chat and nothing in there should be taken as legal advice no matter the speaker’s credentials.

**Comment by marclaporte** [comment]
Just bumping this. The sooner we decide, the sooner we start, the sooner it will be done :-)

**Comment by marclaporte** [comment]
This issue was opened 9 years ago.
